File: #241_Different_Ways_to_Add_Parentheses/Solution.py
class Solution(object):
	# @Param input: String
	# @Return List[int]
	def diffWaysToCompute(self, input):
		if not input:
			return []
		re = []
		for i in range(len(input)):
			if input[i] == '+' or input[i] == '-' or input[i] == '*':
				left = self.diffWaysToCompute(input[0:i])
				right = self.diffWaysToCompute(input[i + 1:])
				for lnum in left:
					for rnum in right:
						if input[i] == '+':
							re.append(lnum + rnum)
						elif input[i] == '-':
							re.append(lnum - rnum)
						else:
							re.append(lnum * rnum)
		if not re:
			re.append(int(input))
		return re

sol = Solution()
print sol.diffWaysToCompute('2-1-1')

File: #241_Different_Ways_to_Add_Parentheses/Solution.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
	vector<int> diffWaysToCompute(string input) {
		vector<int> re;
		if (input.empty()) {
			return re;
		}
		for (int i = 0; i < input.size(); i++) {
			if (input[i] == '+' || input[i] == '-' || input[i] == '*') {
				vector<int> left = diffWaysToCompute(input.substr(0, i));
				vector<int> right = diffWaysToCompute(input.substr(i + 1, input.size() - i - 1));
				for (int l : left) {
					for (int r : right) {
						if (input[i] == '+') re.push_back(l + r);
						else if (input[i] == '-') re.push_back(l - r);
						else re.push_back(l * r);
					}
				}
			}
		}
		if (re.size() == 0) re.push_back(stoi(input));
		return re;
	}
};

int main(int argc, char const *argv[])
{
	Solution sol;
	vector<int> re = sol.diffWaysToCompute("1+1");
	for (int i : re) {
		printf("%d\n", i);
	}
	return 0;
}